### ISSUE TITLE
migrate error when deploy

### ISSUE THREAD

**Issue by Casear** [bitwalker/distillery#266]
I try to deploy phoenix from Jenkins.

According [migrations](https://github.com/bitwalker/distillery/blob/master/docs/Running%20Migrations.md) article,
I created file  "lib/tasks.ex".
```
defmodule App.ReleaseTasks do

  @start_apps [
    :postgrex,
    :ecto
  ]
  
  @app [
    :app
  ]

  @repos [
    App.Repo
  ]

  def seed do
    IO.puts "Loading app.."
    # Load the code for myapp, but don't start it
    :ok = Application.load(:app)

    IO.puts "Starting dependencies.."
    # Start apps necessary for executing migrations
    Enum.each(@start_apps, &Application.ensure_all_started/1)

    # Start the Repo(s) for myapp
    IO.puts "Starting repos.."
    Enum.each(@repos, &(&1.start_link(pool_size: 1)))

    # Run migrations
    Enum.each(@app, &run_migrations_for/1)

    # Run the seed script if it exists
    seed_script = Path.join([priv_dir(:App), "repo", "seeds.exs"])
    if File.exists?(seed_script) do
      IO.puts "Running seed script.."
      Code.eval_file(seed_script)
    end

    # Signal shutdown
    IO.puts "Success!"
    :init.stop()
  end

  def priv_dir(app), do: "#{:code.priv_dir(app)}"

  defp run_migrations_for(app) do
    IO.puts "Running migrations for #{app}"
    Ecto.Migrator.run(App.Repo, migrations_path(app), :up, all: true)
  end

  defp migrations_path(app), do: Path.join([priv_dir(app), "repo", "migrations"])
  defp seed_path(app), do: Path.join([priv_dir(app), "repo", "seeds.exs"])

end

```
when I executed it
```
bin/app command Elixir.App.ReleaseTasks seed
```
I got this error
```
{"init terminating in do_boot",{undef,[{'Elixir.App.ReleaseTasks',seed,[],[]},{init,start_em,1,[]},{init,do_boot,3,[]}]}}
init terminating in do_boot ()
```

Then I put this on rel/config.exs
```
release :app do
  ...
  set commands: [
    "migrate": "rel/commands/migrate.sh"
  ]
end
```
When I executed "bin/app migrate"
It only show "Service Control" description.

Did I do something wrong?

**Comment by bitwalker** [comment]
Did you create the `rel/commands/migrate.sh` script? What version of Distillery are you using? The issue appears to be that it's not finding the module/function (hence the `undef` error), which means either the code isn't on the path, or wasn't compiled into the release for some reason. If you can tell me what version you are running, I can try to reproduce.

**Comment by Casear** [comment]
My distillery version is '1.4.0'.
I have created ```rel/commands/migrate.sh```.
but ```bin/app migrate``` is still not working.

**Comment by bitwalker** [comment]
I'll take a look this evening

**Comment by harmon25** [comment]
i am having the same issue - it appears in the build binary help, under `Custom Commands` it references the wrong path to the migrate script:

`/opt/app/releases/0.0.1/commands/migrate.sh`

Do not believe that is where the migrate script is located - so when running `bin/app migrate` it cannot find the shell script.

**Comment by Casear** [comment]
My most difficult problem is ```undef```.
I need to deploy migration to production.

**Comment by bitwalker** [comment]
I cannot reproduce on OSX. Can one of you try to `chmod +x path/to/migrate.sh`? The script should be under `<deploy_dir>/releases/<version>/commands`. On OSX, the file is not executable, but it still works - I wonder if different shell implementations care about that.

**Comment by harmon25** [comment]
hmm might have had something to do with line endings or something - was creating the file on windows. 
Seems to be working for me now though!

**Comment by harmon25** [comment]
To add to my above comment - very sure this is related to building docker images from a Windows system - Docker for Windows will alter file permissions, likely removing execute from the custom command script.

Will test a bit more to verify...

**Comment by bitwalker** [comment]
If you find out what caused it, and if there is anything I can do on Distillery's end to handle it, please let me know! I'll close this for the time being though.

**Comment by harmon25** [comment]
Thanks for the input @bitwalker!

Another reason for this is when the database configuration fails to connect during a custom command.
Seems to squelch the error a bit, not easy to tell that:
```
{"init terminating in do_boot",{undef,[{'Elixir.App.ReleaseTasks',seed,[],[]},{init,start_em,1,[]},{init,do_boot,3,[]}]}}
init terminating in do_boot ()
```
Is the equivalent to:
```
[error] Postgrex.Protocol (#PID<0.2055.0>) failed to connect: ** (DBConnection.ConnectionError) tcp connect (database.url:5432): timeout
```

Not sure if that error message can be relayed to the shell during a release migration that fails to connect/times out - would have made this much easier to debug...

Likely the cause of this is actually a *no match error*, my release task contains this line:

```elixir
   # Start the Repo(s) for myapp
    IO.puts "Starting repos.."
    {:ok, pid} = App.Repo.start_link(pool_size: 1) #if DB cannot connect this would not return {:ok, _}  
```
It is likely that `start_link` function call that is resulting in:
```
{"init terminating in do_boot".....
```

Something to consider when creating release tasks - need to make sure they don't crash, or else may be difficult to debug!

**Comment by bitwalker** [comment]
The error is stating that `App.ReleaseTasks.seed` is not defined. If `seed` takes one or more arguments, that is likely the reason; it can also happen if the module or function name is typo'd, but that doesn't look like the problem there.

I would like to fix up those error messages, unfortunately those `ini terminating` messages occur because the VM doesn't even fully boot (usually because an application fails to start for some reason). One thing I thought about was storing that error message and parsing it with my release utils escript to pretty print it - but I haven't gone down that road yet - the only thing worse than a cryptic error is no error message at all :(

**Comment by siepet** [comment]
Okay, same error here: I defined a task inside `lib/tasks/release_tasks.exs`, added migrate.sh file that runs this command `Elixir.MyApp.ReleaseTasks migrate` yet it keeps throwing undef as above. Version 1.4.1

**Comment by bitwalker** [comment]
@siepet Did you define it as an `.exs` file? It won't be compiled if it's `.exs` rather than `.ex`

**Comment by joaquimadraz** [comment]
@siepet I had this problem too but in a different scenario.
Like @bitwalker said your `release_tasks.exs` is not being compiled hence the:
```
{"init terminating in do_boot",{undef,[{'Elixir.App.ReleaseTasks',seed,[],[]},{init,start_em,1,[]},{init,do_boot,3,[]}]}}
init terminating in do_boot ()
```

In my case I have an umbrella app and I just created the `lib` folder on the root or the umbrella app.
That folder is not on the path to be compiled so I was having the same error when running the `command`.

To fix it, I just moved the file inside of one of the apps under the umbrella so it gets compiled 👍
